[SYCL] Remove SYCL_DEVICE_FILTER support

uditagarwal97 opened the pull request:
**Overview**
This PR removes support for SYCL_DEVICE_FILTER, and refactors `Basic/diagnostics/device-check.cpp` to use ONEAPI_DEVICE_FILTER instead of SYCL_DEVICE_TYPE, as the latter is depreciated.

sarnex reviewed (approved):
ESIMD changes (sycl/test/invoke_simd/invoke_simd.cpp only) LGTM

uditagarwal97 commented:
@steffenlarsen I've addressed your reviews from https://github.com/intel/llvm/pull/12007 in this PR. Since Euna is inactive, I'll be taking forward this effort (that's why created a separate PR).

aelovikov-intel commented on sycl/source/detail/config.cpp:
```diff
@@ -175,8 +176,21 @@ getSyclDeviceTypeMap() {
+// Array is used by SYCL_DEVICE_ALLOWLIST and ONEAPI_DEVICE_SELECTOR.
+const std::array<std::pair<std::string, backend>, 7> &getSyclBeMap() {
+  static const std::array<std::pair<std::string, backend>, 7> SyclBeMap = {
+      {{"host", backend::host},
```
Is comment on line 165 applicable here as well?

uditagarwal97 replied:
Yes. I'll add the comment.

aelovikov-intel commented on sycl/source/detail/filter_selector_impl.cpp:
```diff
@@ -129,6 +131,7 @@ int filter_selector_impl::operator()(const device &Dev) const {
       else
         DeviceTypeOK = (DT == Filter.DeviceType);
     }
+#ifndef __INTEL_PREVIEW_BREAKING_CHANGES
```
Can you clarify why this code relates only to one path and not the other? It's not immediately clear from the variable names below...

uditagarwal97 replied:
I've fixed this. Apparently, removing this branch was changing the functionality of filter selector. So, instead I've now added a variable "MatchesSeen" in `ods_target` to keep track of the devices that matches this filter.

aelovikov-intel commented on sycl/source/detail/config.cpp:
```diff
@@ -161,8 +161,9 @@ void dumpConfig() {
-// Array is used by SYCL_DEVICE_FILTER and SYCL_DEVICE_ALLOWLIST and
-// ONEAPI_DEVICE_SELECTOR
+// Array is used by SYCL_DEVICE_ALLOWLIST and ONEAPI_DEVICE_SELECTOR.
+// TODO: host device type will be removed once sycl_ext_oneapi_filter_selector
+// is removed.
```
Do you know if we should remove it under the same task (even though in a separate PR)?

uditagarwal97 replied:
AFAIK, from https://github.com/intel/llvm/pull/12007#discussion_r1407724969, we should perhaps first depreciate sycl_ext_oneapi_filter_selector and then, eventually remove or replace it.  I think it would be better to do that in a separate task as it would require more discussion and it is loosely related to the current task of removing depreciated variables.

aelovikov-intel commented on sycl/test-e2e/DeprecatedFeatures/deprecated_sycl_device_filter.cpp:
```diff
@@ -1,6 +1,6 @@
-// RUN: %{build} -o %t.out
-// RUN: env SYCL_DEVICE_FILTER='*' %{run-unfiltered-devices} %t.out &> %t.log
-// RUN: FileCheck %s < %t.log
+// RUN: %if !preview-breaking-changes-supported %{ %{build} -o %t.out %}
```
I don't think we need any changes in the tests using old filter - it's the default still.

gmlueck commented on sycl/source/detail/filter_selector_impl.cpp:
```diff
@@ -129,6 +131,7 @@ int filter_selector_impl::operator()(const device &Dev) const {
       else
         DeviceTypeOK = (DT == Filter.DeviceType);
     }
+#ifndef __INTEL_PREVIEW_BREAKING_CHANGES
```
Are these `#ifdef` lines changing the behavior of the `filter_selector`?  I'm not sure we want to do that, and I didn't see any changes to the "sycl_ext_oneapi_filter_selector" extension specification.

uditagarwal97 replied:
Thanks, I've fixed this. It was indeed changing the behavior of filter selector.

kbenzie reviewed (approved):
UR LGTM

AlexeySachkov commented on sycl/source/detail/config.cpp:
```diff
@@ -175,8 +176,21 @@ getSyclDeviceTypeMap() {
 
-// Array is used by SYCL_DEVICE_FILTER and SYCL_DEVICE_ALLOWLIST and
-// ONEAPI_DEVICE_SELECTOR
+#ifdef __INTEL_PREVIEW_BREAKING_CHANGES
```
Can we just `#ifdef` a single line with `esimd_emulator` and use `auto` return type to unify those two versions?

AlexeySachkov commented on sycl/source/detail/pi.cpp:
```diff
@@ -292,16 +292,22 @@ std::vector<std::pair<std::string, backend>> findPlugins() {
                           "ONEAPI_DEVICE_SELECTOR cannot be used in "
                           "conjunction with SYCL_DEVICE_FILTER");
   } else if (!FilterList && !OdsTargetList) {
+#else
```
Does it make sense to refactor this right away, so it looks cleaner and it will be easier to clean it up later?

My suggestion is to have the following:
```c++
  ods_target_list *OdsTargetList = SYCLConfig<ONEAPI_DEVICE_SELECTOR>::get();
#ifndef __INTEL_PREVIEW_BREAKING_CHANGES
  device_filter_list *FilterList = SYCLConfig<SYCL_DEVICE_FILTER>::get();
  if (OdsTargetList && FilterList) {
    throw ...
  }
  if (FilterList) {
    // existing "else if (FilterList)" branch content
   return PluginNames; // new line to have early exit for simplifying rest of the code
  }
#endif
  if (!OdsTargetList) {
    // we will now have early exit above, so we want enter this branch if FilterList is set
    // existing code for the corresponding branch
  } else {
    // suggested code for the corresponding branch (existing + esimd emulator  handling removal under ifdef)
  }
```

This way, there will be only one huge `#ifdef` section, which can easily be removed later.

uditagarwal97 replied:
Thanks. I've fixed that.

AlexeySachkov reviewed (approved):
`doc` changes LGTM

bader commented on sycl/test-e2e/DeprecatedFeatures/sycl_device_filter.cpp:
```diff
@@ -2,6 +2,7 @@
 // RUN: %if any-device-is-cpu %{ env SYCL_DEVICE_FILTER=cpu %{run-unfiltered-devices} %t.out %}
 // RUN: %if any-device-is-gpu %{ env SYCL_DEVICE_FILTER=gpu %{run-unfiltered-devices} %t.out %}
 // RUN: %if any-device-is-acc %{ env SYCL_DEVICE_FILTER=acc %{run-unfiltered-devices} %t.out %}
+// TODO: Remove this test once SYCL DEVICE FILTER is removed.
```
```suggestion
// TODO: Remove this test once SYCL_DEVICE_FILTER is removed.
```
Let's use the variable name just to make it easier to find in the future.

uditagarwal97 replied:
Sure

bader reviewed (commented):
A couple of minor comments.

I suggest updating PR title and description:
"[SYCL] Remove depreciated environment variables" -> "[SYCL] Remove SYCL_DEVICE_FILTER support".

90+% of this patch related to just one environment variable.

>Same as https://github.com/intel/llvm/pull/12007with additional changes.

> Overview

NOTE: PR description will be a git log message when merged and I don't think we need this this information there. I would move this information from PR description to PR comments.

> This PR removes depreciated environment variables, namely SYCL_BE, SYCL_DEVICE_FILTER, and SYCL_DEVICE_TYPE, along with depreciated esimd_emulator backend.

1. I don't see any instance of `SYCL_BE` in this PR.
2. SYCL_DEVICE_TYPE is used only in one test which just checks runtime emits diagnostics when this variable is set.
3. `esimd_emulator` is not an environment variable, so I would move related changes to a separate PR.

gmlueck reviewed (approved):
Doc and spec changes LGTM.

uditagarwal97 commented:
> A couple of minor comments.
> 
> I suggest updating PR title and description: "[SYCL] Remove depreciated environment variables" -> "[SYCL] Remove SYCL_DEVICE_FILTER support".
> 
> 90+% of this patch related to just one environment variable.
> 
> > Same as https://github.com/intel/llvm/pull/12007with additional changes.
> 
> > Overview
> 
> NOTE: PR description will be a git log message when merged and I don't think we need this this information there. I would move this information from PR description to PR comments.
> 
> > This PR removes depreciated environment variables, namely SYCL_BE, SYCL_DEVICE_FILTER, and SYCL_DEVICE_TYPE, along with depreciated esimd_emulator backend.
> 
> 1. I don't see any instance of `SYCL_BE` in this PR.
> 2. SYCL_DEVICE_TYPE is used only in one test which just checks runtime emits diagnostics when this variable is set.
> 3. `esimd_emulator` is not an environment variable, so I would move related changes to a separate PR.

Thanks, Alexey. I've made most of the suggested changes. Regarding moving `esimd_emulator` related changes to a separate PR, my concern is that `esimd_emulator` is supported only with SYCL_DEVICE_FILTER, not with ONEAPI_DEVICE_SELECTOR. So, the stuff I am migrating from SYCL_DEVICE_FILTER to ONEAPI_DEVICE_SELECTOR (like filter_selector) might fail if we don't remove the `esimd_emulator` support right away. The converse of this, i.e. removing `esimd_emulator` support before merging this PR, would also be incorrect as it is still valid for SYCL_DEVICE_FILTER.

bader reviewed (approved):
Thanks!

uditagarwal97 commented:
Converted this PR to draft to prevent accidental merges. I'm trying to separate out esimd_emulator related changes from this PR to https://github.com/intel/llvm/pull/12428. Once it gets merged, I'll refactor and open this PR again.

uditagarwal97 commented:
@aelovikov-intel @bso-intel could you review this PR again? The only change made since the last review is the separation of ESIMD related changes in https://github.com/intel/llvm/pull/12428

aelovikov-intel commented:
Post commit failures on Arc GPU:

```
Failed Tests (2):
  SYCL :: Basic/get_backend.cpp
  SYCL :: OneapiDeviceSelector/level_zero_top.cpp
```

The first is kind of known, see https://github.com/intel/llvm/pull/12463#issuecomment-1909135249.
The second is new to me, but looks passing in the previous/next jobs, so possibly unrelated.
